arith/field_hashers/src/poseidon/impls.rs

```diff
@@ -0,0 +1,210 @@
+use std::{
+    fmt::Debug,
+    ops::{Add, AddAssign, Mul, MulAssign},
+};
+
+use arith::{Field, FieldForECC};
+use tiny_keccak::{Hasher, Keccak};
+
+use super::compile_time::compile_time_sbox_alpha;
+use crate::FiatShamirFieldHasher;
+
+pub trait PoseidonStateTrait:
+    Sized
+    + Default
+    + Clone
+    + Debug
+    + Add<Self, Output = Self>
+    + for<'a> Add<&'a Self, Output = Self>
+    + AddAssign<Self>
+    + for<'a> AddAssign<&'a Self>
+    + Mul<Self, Output = Self>
+    + for<'a> Mul<&'a Self, Output = Self>
+    + MulAssign<Self>
+    + for<'a> MulAssign<&'a Self>
+{
+    /// ElemT is the element (base field) type constructing the poseidon state instance.
+    type ElemT: FieldForECC;
+
+    /// SBOX_POW is the lowest exponential pow for poseidon sbox.
+    const SBOX_POW: usize = compile_time_sbox_alpha::<Self::ElemT>();
```

Comment:
shall we hard code this in each field's implementation?

Reply:
fair point - I removed the compile time alpha derivation, and leave the trait const being empty and up to implementations and instantiations in the latest commit.